internal/config/config.go

```diff
@@ -0,0 +1,167 @@
+limitations under the License.
+*/
+
+package config
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/alecthomas/kingpin/v2"
+	"gopkg.in/yaml.v3"
+)
+
+// Config represents the complete application configuration
+type (
+	Log struct {
+		Level  string `yaml:"level"`
+		Format string `yaml:"format"`
+	}
+
+	Config struct {
+		Log Log `yaml:"log"`
+	}
+)
+
+const (
+	// Flags
+	LogLevelFlag  = "log.level"
+	LogFormatFlag = "log.format"
+)
+
+// DefaultConfig returns a Config with default values
+func DefaultConfig() *Config {
+	cfg := &Config{}
+
+	cfg.Log.Level = "info"
+	cfg.Log.Format = "text"
+
+	return cfg
+}
+
+// Load loads configuration from an io.Reader
+func Load(r io.Reader) (*Config, error) {
+	cfg := DefaultConfig()
+
+	data, err := io.ReadAll(r)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read config: %w", err)
+	}
+
+	if err := yaml.Unmarshal(data, cfg); err != nil {
+		return nil, fmt.Errorf("failed to parse config: %w", err)
+	}
+	cfg.sanitize()
+
+	if err := cfg.Validate(); err != nil {
+		return nil, err
+	}
+
+	return cfg, nil
+}
+
+// FromFile loads configuration from a file
+func FromFile(filePath string) (*Config, error) {
+	file, err := os.Open(filePath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open config file: %w", err)
+	}
+	defer file.Close()
+
+	return Load(file)
+}
+
+type ConfigUpdaterFn func(*Config) error
+
+// RegisterFlags registers command-line flags with kingpin app
+// and returns ConfigUpdaterFn that updates the config from parsed flags
+// as command line arguments override config file settings
+func RegisterFlags(app *kingpin.Application) ConfigUpdaterFn {
+	// track flags that were explicitly set
+	flagsSet := map[string]bool{}
+
+	app.PreAction(func(ctx *kingpin.ParseContext) error {
+		// Clear the map in case this function is called multiple times
+		flagsSet = map[string]bool{}
+
+		for _, element := range ctx.Elements {
+			if flag, ok := element.Clause.(*kingpin.FlagClause); ok && element.Value != nil {
+				flagsSet[flag.Model().Name] = true
+			}
+		}
+		return nil
+	})
+
+	// Logging
+	logLevel := app.Flag(LogLevelFlag, "Logging level: debug, info, warn, error").Default("info").Enum("debug", "info", "warn", "error")
+	logFormat := app.Flag(LogFormatFlag, "Logging format: text or json").Default("text").Enum("text", "json")
+
+	return func(cfg *Config) error {
+
+		// Logging settings
+		if flagsSet[LogLevelFlag] {
+			cfg.Log.Level = *logLevel
+		}
+
+		if flagsSet[LogFormatFlag] {
+			cfg.Log.Format = *logFormat
+		}
+
+		cfg.sanitize()
+		return cfg.Validate()
+	}
+}
+
+func (c *Config) sanitize() {
+	c.Log.Level = strings.TrimSpace(c.Log.Level)
+	c.Log.Format = strings.TrimSpace(c.Log.Format)
+}
+
+// Validate checks for configuration errors
+func (c *Config) Validate() error {
+	var errs []string
+	{ // log level
+
+		validLogLevels := map[string]bool{
+			"debug": true,
+			"info":  true,
+			"warn":  true,
+			"error": true,
```

Comment:
doesn't kingpin already perform the same validation?

Reply:
but we also want the config read from yaml to be validated as well.